src/service/url-replacements-impl.js

```diff
@@ -598,6 +777,69 @@ export class UrlReplacements {
     // FOO_BAR(arg1,arg2)
     return new RegExp('\\$?(' + all + ')(?:\\(([0-9a-zA-Z-_.,]+)\\))?', 'g');
   }
+
+  /**
+   * Determines the offset of an element's shadowRoot host if it has one,
+   * otherwise return top/left 0.
+   * @param {EventTarget} element
+   * @return {!{top: number, left: number}}
+   * @private
+   */
+  getShadowHostOffset_(element) {
+    if (element) {
+      const shadowRoot = closest(element, parent => {
+        return parent.nodeType == 11;
+      });
+      if (shadowRoot && shadowRoot.host) {
+        // Can we guarantee offsetTop/Left are correct values without
+        // forcing a promise measure?  Given we cannot wait, what values would
+        // we expect here?
+        return {top: shadowRoot.host./*REVIEW*/offsetTop,
+          left: shadowRoot.host./*REVIEW*/offsetLeft};
+      }
+    }
+    return {top: 0, left: 0};
+  }
+
+  /**
+   * If event target is anchor, expand href synchronously via expandSync.
+   * Original href is stored in data-amp-orig-href so that subsequent calls
+   * will generate newest href value.
+   * @param {Event} evt click event
+   * @return {string} expanded url from event target href
+   * @private
+   */
+  getExpandAnchorHref_(evt) {
+    if (!evt || !evt.target || evt.target.tagName !== 'A') {
```

Comment:
Do you need to handle cases where the clicked element is a child of the `a` tag?